traits/trait_types.py

```diff
@@ -3417,7 +3420,7 @@ def validate(self, object, name, value):
             self.validate_failed(object, name, value)
         else:
             result = self.default_value
-            if isinstance(result, _InstanceArgs):
+            if self.default_value_type == DefaultValue.callable_and_args:
```

Comment:
This line isn't exercised by tests; I tried to fix that, but found that it's actually rather hard to exercise this code path (in most cases we're using `Instance` rather than `BaseInstance`, with its fast validation; moreover, the validation for `adapt=='default'` doesn't seem to behave the way it should). I'm planning another PR to clean this up.

Reply:
That PR was created and merged into this one.